### PR TITLE
Add TreeEnsembleClassifier node

### PR DESCRIPTION
**Pull request by tlane-rbx**
Adds [TreeEnsembleClassifier](https://onnx.ai/onnx/operators/onnx_aionnxml_TreeEnsembleClassifier.html) node support. There aren't any backend tests for this node, likely due to it being deprecated. This can probably be easily adapted to support the newer [TreeEnsemble](https://onnx.ai/onnx/operators/onnx_aionnxml_TreeEnsemble.html) node which has backend tests available in newer versions of the onnx submodule than is currently available here.

I'll try to update this PR with tests if I get time soon. The implementation isn't complete (only supports one post_transform function) but I assume some implementation is better than none.

### CODE REVIEW

**Comment by kraiskil** [comment]
Thanks.
I see nothing wrong with this. But then, this algorithm is not familiar to me. Really unfortunate that ONNX doesn't have any tests...

I agree, a partial implementation is better than none. And as long as it throws an `ERROR()` when a non-implemented algorithm is requested, its all good.

I'll wait a bit before merging, just in case you find time to add at least one test for this. But it is worth merging even without one.

Out of curiosity, where did you conjure up a TreeEnsembleClassifier? I haven't bumped into ONNX-ML nodes anywhere.

**Comment by tlane-rbx** [comment]
I was working with a model that was originally converted from sklearn which made use of TreeEnsembleClassifier. It's probably a pretty niche node for ONNX users.

File: src/nodes/treeensembleclassifier.cc [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	if( mode == "LEAF" )
+		ERROR("Attempt to convert LEAF mode in TreeEnsembleClassifier");
+	ERROR("Unknown mode in TreeEnsembleClassifier");
+};
+
+std::unordered_map<int64_t, TreeEnsembleClassifier::Tree> TreeEnsembleClassifier::generateTreeMap() const
+{
+	std::unordered_map<int64_t, Tree> trees;
+
+	for( int i = 0; i < node_tree_ids.size(); ++i )
```
Comment (kraiskil) [inline src/nodes/treeensembleclassifier.cc:98]:
```suggestion
	for( unsigned i = 0; i < node_tree_ids.size(); ++i )
```

File: src/nodes/treeensembleclassifier.cc [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+	{
+		int64_t treeId = node_tree_ids[i];
+		int64_t nodeId = node_node_ids[i];
+		trees[treeId][nodeId] = std::make_shared<TreeNode>();
+	}
+
+	for( int i = 0; i < leaf_tree_ids.size(); ++i )
```
Comment (kraiskil) [inline src/nodes/treeensembleclassifier.cc:105]:
```suggestion
	for( unsigned i = 0; i < leaf_tree_ids.size(); ++i )
```

File: src/nodes/treeensembleclassifier.cc [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		int64_t treeId = leaf_tree_ids[i];
+		int64_t leafId = leaf_node_ids[i];
+		trees[treeId][leafId]->is_leaf = true;
+		trees[treeId][leafId]->class_ids.push_back(leaf_class_ids[i]);
+		trees[treeId][leafId]->weights.push_back(leaf_node_weights[i]);
+	}
+
+	// This part must be done separate from creating the TreeNodes
+	// so they all exist before connections start being made.
+	for( int i = 0; i < node_tree_ids.size(); ++i )
```
Comment (kraiskil) [inline src/nodes/treeensembleclassifier.cc:116]:
```suggestion
	for( unsigned i = 0; i < node_tree_ids.size(); ++i )
```

File: src/nodes/treeensembleclassifier.cc [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+		else if(post_transform == "LOGISTIC")
+		{
+			dst << "1.0f / (1.0f + exp(result_class_" << class_id << "));" << std::endl;
+		}
+	}
+
+	if( unused_class_id != -1 )
+	{
+		INDT_1 << "probabilities[" << unused_class_id << "] = 1.0f - (";
+		for( int i = 0; i < class_ids.size(); ++i )
```
Comment (kraiskil) [inline src/nodes/treeensembleclassifier.cc:216]:
```suggestion
		for( unsigned i = 0; i < class_ids.size(); ++i )
```

**Review by kraiskil — COMMENTED** [review]
Thanks for the test, this looks really complete now.

I tried to compile locally, and the `-Wall -Werror` flags caught a few warnings. If you could commit those smalle changes, then this is ready to merge in.

**Comment by kraiskil** [comment]
Unrelated to this, but seems I am a newbie to GitHub :)

The 'commit changes' button for the changes I suggest is _not_ grayed out for me.
Where would the commit end up in, if I click it? I shouldn't have access to commit to the branch in your fork, right? I don't dare click it now... But if the 'commit changes' button is greyed out for you, maybe I must.

**Comment by tlane-rbx** [comment]
> Unrelated to this, but seems I am a newbie to GitHub :)
> 
> The 'commit changes' button for the changes I suggest is _not_ grayed out for me. Where would the commit end up in, if I click it? I shouldn't have access to commit to the branch in your fork, right? I don't dare click it now... But if the 'commit changes' button is greyed out for you, maybe I must.

It's available to both of us because I clicked a box that says "Allow edits by maintainers". It could catch people off guard if they don't expect it and try to commit without pulling first. So probably _generally_ best not to click it yourself if only to allow potential discussion with the contributor regarding the change.

File: src/nodes/treeensembleclassifier.cc (hunk 1 above)
Comment (tlane-rbx) [reply src/nodes/treeensembleclassifier.cc]: Good catches. I wonder why these didn't get flagged on my local builds. Maybe my compiler thinks these are fine or there is a CMake issue.

File: src/nodes/treeensembleclassifier.cc [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	if( unused_class_id != -1 )
+	{
+		INDT_1 << "probabilities[" << unused_class_id << "] = 1.0f - (";
+		for( size_t i = 0; i < class_ids.size(); ++i )
+		{
+			if( class_ids[i] == unused_class_id )
+			{
+				continue;
+			}
+			dst << "probabilities[" << class_ids[i] << std::string("]") + (i < class_ids.size() - 1 && unused_class_id != i + 1 ? " + " : "");
```
Comment (kraiskil) [inline src/nodes/treeensembleclassifier.cc:222]: One more sign mismatch
```suggestion
			dst << "probabilities[" << class_ids[i] << std::string("]") + (i < class_ids.size() - 1 && unused_class_id != (int)i + 1 ? " + " : "");
```

File: src/nodes/treeensembleclassifier.cc (hunk 1 above)
Comment (kraiskil) [reply src/nodes/treeensembleclassifier.cc]: I think I add `-Wall -Werror` to the Debug build only, but I'd guess you are building that one too. Maybe someone in the compilier project thought the warning is pretty silly (which it actually is, that should not cause problems as long as the tree is not insanely big). FWIW, I'm on gcc 9.4.

Mostly these are annoyance - I want to keep the build warning free just to not become unsensitized to warnings :)

File: src/nodes/treeensembleclassifier.cc (hunk 5 above)
Comment (kraiskil) [reply src/nodes/treeensembleclassifier.cc:222]: Sorry about this. Now _I_ don't understand how I didn't catch this yesterday :/

Comment (tlane-rbx) [reply src/nodes/treeensembleclassifier.cc]: I introduced that one by fixing an issue from yesterday

File: src/nodes/treeensembleclassifier.cc (hunk 1 above)
Comment (tlane-rbx) [reply src/nodes/treeensembleclassifier.cc]: I'm on Clang 15. My `compile_commands.json` seems to have `-Wall -Werror` also. Very weird, I'd expect to see the same warning. Perhaps worth looking into.

**Pull request merged**